### PR TITLE
New field types :yum::flushed: [WIP]

### LINKED ISSUE #641: Enable ability to apply multiple types to metabase fields
Right now each Field in a table can only have a single type, which is very limiting.  There is a very real reason to be able to apply multiple types to each field.

Possibly move metabase types to an M2M tag

We have a very strict set of types for entities and fields. We should make them a tag based system to allow for a hierarchical ontology.

We would use this to allow multiple levels of semantic information to be carried through the system. EG. we can have things marked as both dimensions, metrics, as well as an underlying type e.g. datetime vs category (i.e. low cardinality int or string)

### PR DESCRIPTION
**Pull request by camsaul**
Proposed new Field typing system. Replace our legacy [Django-derived](https://docs.djangoproject.com/en/1.9/ref/models/fields/#field-types) field type system with something extensible that will meet our needs moving forward. This will allow us to resolve #641, #708, #824, #1324, #1367, #1522, #1576, #1687, #1698, #1723, #1858, #1889, #1891, and #1929.

The basic change here is replacing the flat type system with a hierarchical one. `:BigIntegerField` becomes `:number.integer.big`, `:DecimalField` becomes `:number.float.decimal`, and so forth. This allows individual drivers to extend the type hierarchy when appropriate, e.g. `:text.mongo-bson-id`. The backend knows new types like these derive from base types via [`extend`](http://clojuredocs.org/clojure.core/extend)/[`isa?`](http://clojuredocs.org/clojure.core/isa_q) while the frontend can check whether the type starts with a given string (e.g. _starts with_ `:type/datetime`). This simplifies complicated logic such as the logic involved in determining whether something is a datetime field.

The other big change is unifying the valid types for `base_type` and `special_type`. `special_type` becomes more of a "virtual type" or "effective type" and specifies how we should _use_ a Field, while `base_type` becomes more of a "native type" that specifies _how the Field is stored internally_ in a database. Since `pk`-ness and `category`-ness aren't really related to the type of the Field I've separated those out into separate boolean columns.

[WIP]

### CODE REVIEW

**Comment by agilliland** [comment]
Cool.  This is looking pretty neat.  

Looking at this briefly one of the questions in my mind is if we'll run into issues where combining the type of the data (text, number, decimal, boolean) with the parsing logic (timestamp.milliseconds) with the potential special meaning (geo.zip) will make for redundancies.  It seems like there would be possible ways where zip code could be represented as text or a number.  Or where percentage could be either a decimal or whole number.  Anyways, something we can think about.

File: resources/migrations/030_add_new_field_types.yaml [hunk 1]
```diff
@@ -0,0 +1,130 @@
...
+            sql: UPDATE metabase_field SET new_base_type = 'type/text.uuid' WHERE base_type ='UUIDField'
+        - sql:
+            sql: UPDATE metabase_field SET new_base_type = 'type/*' WHERE base_type ='UnknownField'
+        - dropColumn:
+            tableName: metabase_field
+            columnName: base_type
+        - renameColumn:
+            tableName: metabase_field
+            oldColumnName: new_base_type
+            newColumnName: base_type
```
Comment (agilliland) [inline resources/migrations/030_add_new_field_types.yaml]: renaming columns is something we should never need to do in migration scripts.  it's always easier to just create a new column, copy the necessary data over, then delete the old column whenever we want.

and i think we should avoid having any drop column statements until we get to major version changes.  the reason being that dropping a column creates a 100% guarantee that you can't revert from one version to a previous one.  Hopefully not something that users never need to do but it's good to keep it possible.

**Comment by agilliland** [comment]
One other question as I'm poking around ... where would we capture the database specific type information?  so for example of we had a Field that was a postgres `inet` type column, how do we capture that?

**Comment by camsaul** [comment]
For something like zip I see the column having a base type of either `number.integer` or `text` and a special type of `number.integer.geo.zip`. The special type would mostly inform the UI/chart options -- it would be how we "treat" the column, regardless of how it's actually stored. The base type would be used so we could do text -> integer conversions or whatever else as needed.

So the drop column / rename column stuff was just what I found to be easiest for converting the old style to the new one. Keeping the `base_type` and `special_type` names makes this huge PR a little more manageable. Otherwise after some initial investigation it would be a huge effort to change all the references in the code base for the columns when most of the time it isn't needed.

Postres `inet` would be something like `text.ip-address.postgres-inet` I think. So we'd display the `text` widget (or even a `text.ip-address` if we made one). But Postgres could treat it specially however it needed

**Comment by camsaul** [comment]
@agilliland take another look at this. I reworked the migration to just keep the existing columns and change everything in place.

This PR is now effectively one big Emacs `projectile-replace`. Base types & special types have a unified type hierarchy and things like `BigIntegerField` become `type/number.integer.big`. In Clojure we determine whether something is of a given type with [`isa?`](http://clojuredocs.org/clojure.core/isa_q); in JS we check [`String.protoype.startsWith()`](https://developer.mozilla.org/en/docs/Web/JavaScript/Reference/Global_Objects/String/startsWith). This allows drivers to extend the type system to suit their needs (e.g. `type/number.integer.big.sql-server-5000-bit-integer`) without having to change code anywhere else in the project.

For now `fk`, `id`, and `category` are still special types instead of being boolean properties of `Field` itself. That's something we'll have to revisit later on.

WIP: I still need to manually test the frontend a bit more and make sure everything is working as intended. But if this PR is looking good I'd like to merge it so I can tackle #641, #708, #824, #1324, #1367, #1522, #1576, #1687, #1698, #1723, #1858, #1889, #1891, and #1929, and hopefully get the majority of them done in time for 15.0

**Comment by tlrobinson** [comment]
`special_type` isn't guaranteed to be not null, is it? If not the we can't just do `col.special_type.startsWith(...)`. We could add an `isa` function to the schema_metadata.js module or something.

File: frontend/src/lib/schema_metadata.js [hunk 2]
```diff
@@ -20,19 +20,19 @@ export const UNKNOWN = "UNKNOWN";
...
+        special: ["type/datetime.unix.milliseconds", "type/datetime.unix.seconds"]
     },
     [NUMBER]: {
-        base: ["IntegerField", "DecimalField", "FloatField", "BigIntegerField"],
+        base: ["type/number.integer", "type/number.float.decimal", "type/number.float", "type/number.integer.big"],
         special: ["number"]
     },
     [STRING]: {
-        base: ["CharField", "TextField"],
+        base: ["type/text", "type/text"],
```
Comment (tlrobinson) [inline frontend/src/lib/schema_metadata.js]: Duplicate

**Comment by camsaul** [comment]
@tlrobinson Yup, good point, I'll add an `isa` function so if we want to tweak how it works down the road we don't need to change logic in a dozen different places

**Comment by agilliland** [comment]
I think this is moving in the right direction but I still think there are a couple things we need to address.  There are a few requirements that are laid out in our design document on GoogleDocs which I want to make sure we are satisfying here:

1. The data_type must be something that can be controlled by the user via the UI.
    * why?  to allow casting where the inferred data_type was wrong.  for example a varchar column has an inferred data_type of `text`, but in this case the user that it only has numbers in it and thus should be treated like a `number.integer` or `number.decimal`, so they make that choice.
2. There should be a place for the database specific column type to be captured and maintained that isn’t editable by the user.
    * why?  in order for a driver to know that casting is necessary it must know that the physical column type differs from the intended data_type the application wants to use.  this is also just a valuable thing for reference when we present the data model to admins in the UI.
3. It must be possible for transformations to be specified.
    * why?  we’ve had several requests for situations where a column needs some degree of parsing/transformation before it’s ready for actual use.  examples: long integer -> unix_timestamp -> datetime or string ‘yyyymmdd’ -> date_parse -> date
    * there are quite a few scenarios where this could be useful, so we should be flexible here.  things like regex matching to pull out part of a value, extracting values from a JSON blob or array, converting an integer column representing currency in cents to a decimal column for currency in dollars, etc.  we should assume there are desired parsing/transformation actions that we don’t know about.

Looking at the current code it doesn't appear to me that we are fully accounting for these requirements yet.  I like the tree based approach for the individual identifiers combined with the use of an `isa?` function so that types can test types at varying levels, so it makes sense to me to continue with that.  What I think still needs some tweaking:

* attaching database specific column type information to the data_type (e.g. `type/number.integer.big.sql-server-5000-bit-integer`) creates a problem where we can't effectively allow users to modify the data_type themselves because it won't actually be a fixed and concise set of choices.  with the database specific column type mixed into the attribute it becomes difficult to deal with.  my recommendation here would be to consider separating out the database specific column type into a separate attribute.
* similarly, combining the data_type with it's transformation (e.g. `datetime.unix.milliseconds`) means that the data_type selection again gets longer because we have to produce a list that is the cross product of all data_types against their allowed transformations.  we could get probably get away with this as-is, but as we look at growing the set of allowed transformations over time this could get unwieldy, so my recommendation would be to consider making the transformation setting a separate attribute.

For the handling of the special_type field I'm fine with the tweaks you are suggesting and I generally agree that updating the underlying system without modifying the options and what they do is a good step and it'll put us in a better place to start making more changes in the next pass.

**Comment by camsaul** [comment]
I think you're misunderstanding it. There is no `data_type` column:

1.  `base_type` is the database-specific type information. That's where we'd set something like `number.integer.big.sql-server-wacky-column-type`. 
2.  `special_type` is the "effective" type chosen by the user or inferred automatically. A user can pick that something is a `datetime` or a `number` even if the underlying type is `text.sql-server`. The driver would then know to do a `text.sql-server-utf-16 -> number` transformation (or perhaps fall back to a `text -> number` transformation). If the user picks `datetime.unix.milliseconds` as the `special_type` the driver can figure out what transformations need to be done based on its `base_type`.

**Comment by agilliland** [comment]
Yeah, and that makes sense.  `data_type` is my own way of referencing the specific way a db column's value is meant to be treated.  But the same limitations still apply from what I said where the code is bundling what feels like several distinct settings/choices together into a single type and I'm concerned that we are trying to do too much with a single attribute.  

So the question is, why is it better to combine the 1) intended value type w/ any (optional) transformations w/ any (optional) distinctions about semantic meaning into a single attribute?

**Comment by camsaul** [comment]
How do you see it working if we have a separate "transformation" type? For something like "Unix Timestamp (milliseconds)" would the user first pick "Parse as Unix Timestamp (milliseconds)" from a "transformations" menu and then "treat as datetime" from a "semantic menu"? That seems redundant since the former implies the latter.

Drivers can figure out how to perform transformations themselves based on the desired semantic type. If we want to add support for transformations like "always lowercase this field and add the prefix XYZ", I consider that a separate issue we can revisit down the road since it doesn't really have anything to do with the semantic meaning of a field.

This PR is just to make our existing type system more extensible so we can add support for new types like "currency" and "Postgres JSON" without having to edit hardcoded mappings in 2 dozen different files.

**Comment by agilliland** [comment]
i am happy to deal with implementing transformations in a later PR, but it’s a known requirement that we need to support, so i bring it up because any changes we make now should account for them in the design.

at this point, here’s what i’m still not comfortable with and would like to hear more justification for:

- putting multiple choices into the special-type attribute.  this feels problematic because it conflates lots of individual options together and makes it harder to handle varied intersections between those choices.  why are we not separating out the choices into distinct attributes?
- having the value casting logic attached to the special-type.  it’s unclear to me why the information about how a field is meant to be used from a value perspective should be coupled together with other choices such as semantic meaning, transformations, formatting, etc.  why would we not place the value type information in a separate attribute?
- having the value type information in multiple attributes.  the current proposal suggests that the code will have to check both special-type and base-type to be sure of what the desired value handling logic is, which feels unnecessary.  from the code you see things like `(or (isa? base-type …) (isa? special-type …))`.  i’d like to understand why it’s not preferred to have a single, known, required attribute where we maintain the desired value type of the field that gets used 100% of the time?
- putting the database specific type at the end of the base_type.  this makes it problematic to use that value for display purposes if we want to put it on the UI somewhere.

i like the mechanism you are using with the isa? function and the type tree which allows for a structured hierarchy of types, so i suggest that we continue to apply that approach but that we should break apart the various pieces of the metadata into more fields which can be independently modified as described in the GoogleDoc.

**Comment by camsaul** [comment]
@agilliland To me it seems like you're criticizing patterns in the existing code for working the way they've always worked, and that you're uncomfortable considering any marginal improvements to the type system unless the entire thing is completely overhauled.

I just wanted to make it easier to fix various type related things in the near future without each of them having to change 2 dozen files like this one had to.

**Comment by salsakran** [comment]
Going to keep talking here and we can move comments if appropriate to a new issue/PR

:+1: on taking a crack at this, @camsaul . We definitely need to rework our metadata ontology, and a refactor is long due.

Regarding the actual choices in the PR:

I'm **very** skitish about anything that leaves heirarchical information in a free string field that has to be parsed to get meaning out of it. Not only does this not let us lean on constraints and relational form to make sure metadata is consistent, but it also makes it hard to point the query builder at our own database and dogfood our own tool. Any data schema decisions we make should absolutely consider ease of use with Metabase itself a constraint.

This also seems like it would make things slightly easier on the backend/driver side, be neutral on metadata heuristics, and **really** messy on the front end side. 

It also conflates semantic and base type into the same field in a semi-predictable fashion.

Strong :+1: on the notion of a heirarchical type structure, though I think we should be much more deliberate about separating out the mapping between underlying bits and "data type" and our semantic model. 

Strong :-1: on the actual format. 

Overall, I like the idea of having these discussions around a tangible PR so let's do more of this.

**Pull request closed without merging**